File: loop_transformations/sources/unroll.2.c
/*
* @@name:	unroll.2
* @@type:	C
* @@operation:	compile
* @@expect:	ct-error
* @@version:	omp_5.1
*/
void illegal_2a(double A[])
{
    #pragma omp for
    #pragma omp unroll full  // ERROR: No loop left after full unrolling.
    for (int i = 0; i < 12; ++i)
        A[i] = 0;
}

void illegal_2b(double A[])
{
    // Loop might be fully unrolled (or a partially unrolled loop
    // replacement). Hence, no canonical for-loop, resulting in
    // non-compliant code. Implementations may suggest adding a
    // "partial" clause.

    #pragma omp for          //        Requires a canonical loop
    #pragma omp unroll       // ERROR: may result in non-compliant code
    for (int i = 0; i < 12; ++i)
        A[i] = 0;
}

void illegal_2c(int n, double A[])
{
    #pragma omp unroll full  // ERROR: Constant iteration count required.
    for (int i = 0; i < n; ++i)
        A[i] = 0;
}

File: loop_transformations/sources/unroll.2.f90
! @@name:       unroll.2
! @@type:       F-free
! @@compilable: yes
! @@linkable:   no
! @@expect:     success
! @@version:    omp_5.1
subroutine illegal_2a(A)
   implicit none
   double precision :: A(*)
   integer :: i

   !$omp do
   !$omp unroll full  !! ERROR: No loop left after full unrolling
   do i = 1,12
      A(i) = 0.0d0
   end do
end subroutine

subroutine illegal_2b(A)
   implicit none
   double precision :: A(*)
   integer :: i

   !! Loop might be fully unrolled (or a partially unrolled loop 
   !! replacement).  Hence, no canonical do-loop will exist, 
   !! resulting in non-compliant code.
   !! Implementations may suggest to adding a "partial" clause.

   !$omp do           !!        Requires a canonical loop
   !$omp unroll       !! ERROR: may result in non-compliant code
   do i = 1,12
      A(i) = 0.0d0
   end do
end subroutine

subroutine illegal_2c(n, A)
   implicit none
   integer          :: i,n
   double precision :: A(*)

   !$omp unroll full  !! Full unroll requires constant iteration count
   do i = 1,n
      A(i) = 0.0d0
   end do
end subroutine
